### PR TITLE
MM-18590 - Custom Admin Settings

### PR DESCRIPTION
**Pull request by marianunez**
#### Summary

Implemented custom admin settings to configure custom attribute that can have name with the associated users and/or groups that it would be applied to.

This first implementation of this UI allows for users to be searched and selected by username. Groups must be added by group id separated by a space (due to limitations by [MM-22071](https://mattermost.atlassian.net/browse/MM-22071))

#### Screenshots

Screen Shot 2020-01-29 at 9 23 17 PM
Screen Shot 2020-01-29 at 9 26 24 PM
Screen Shot 2020-01-29 at 9 27 58 PM

#### Ticket Link
[MM-18590](https://mattermost.atlassian.net/browse/MM-18590)

### CODE REVIEW

File: webapp/.gitignore [hunk 1]
```diff
@@ -1,2 +1,5 @@
 node_modules
 .npminstall
+.vscode/
```
Comment (lieut-data) [inline webapp/.gitignore]: This should be redundant, given that `.gitignore` /should/ recurse to the directory root. Can you confirm?

File: webapp/src/components/admin_settings/users_input/users_input.jsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import debounce from 'lodash/debounce';
+import AsyncSelect from 'react-select/async';
+
+// UsersInput searches and selects user profiles displayed by username.
+// Users prop can handle the user profile object or strings direclty if the user object is not available.
```
Comment (lieut-data) [inline webapp/src/components/admin_settings/users_input/users_input.jsx]:
```suggestion
// Users prop can handle the user profile object or strings directly if the user object is not available.
```

File: webapp/src/components/admin_settings/users_input/users_input.jsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        }).catch(() => {
+            callback([]);
+        });
+    }, 150);
+
+    usersLoader = (term, callback) => {
+        try {
+            this.debouncedSearchProfiles(term, callback);
+        } catch (error) {
+            callback([]);
```
Comment (lieut-data) [inline webapp/src/components/admin_settings/users_input/users_input.jsx]: Should we log something here (and above) to help with debugging if it ever fails?

File: webapp/src/components/admin_settings/custom_attribute_settings.jsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        };
+    }
+
+    initAttributes(attributes) {
+        if (!attributes) {
+            return new Map();
+        }
+
+        // Store the attributes in a map indexed by position
+        const attributesMap = new Map(attributes.map((a, index) => [index, a]));
```
Comment (lieut-data) [inline webapp/src/components/admin_settings/custom_attribute_settings.jsx]: nit: just return without naming at all?

File: webapp/src/components/admin_settings/custom_attribute/custom_attribute.jsx [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        if (!this.props.users || !this.props.users.length) {
+            return;
+        }
+
+        const profiles = await this.props.actions.getProfilesByIds(this.props.users);
+
+        let users = profiles.data;
+
+        if (users.length !== this.props.users.length) {
+            //Check if all ids where returned.
```
Comment (lieut-data) [inline webapp/src/components/admin_settings/custom_attribute/custom_attribute.jsx]:
```suggestion
            // Check if all ids were returned.
```

File: webapp/src/components/admin_settings/custom_attribute/custom_attribute.jsx [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+            //MM-Redux removes the current admin user from the result
```
Comment (lieut-data) [inline webapp/src/components/admin_settings/custom_attribute/custom_attribute.jsx]:
```suggestion
            // mattermost-redux removes the current admin user from the result
```

Might we also link to the offending code in question to help the future issue?

Comment (marianunez) [reply webapp/src/components/admin_settings/custom_attribute/custom_attribute.jsx]: Like a github link ? 

https://github.com/mattermost/mattermost-redux/blob/5f5a8a5007661f6d54533c2b51299748338b5a65/src/actions/users.ts#L340

File: webapp/.gitignore (hunk 1 above)
Comment (marianunez) [reply webapp/.gitignore]: Correct. Removed

**Pull request merged**